### ISSUE TITLE
Workflow Task Failed with Unspecified Cause

### ISSUE THREAD

**Issue by Uria1** [temporalio/sdk-typescript#1115]
## Expected Behavior
A workflow with failed activities should:

1. Change the status to "Failed"
2. Execute code in a `catch` block if the failed activity is awaited from within a `try` block

## Actual Behavior
1. Workflow remains in status "Running" and later times out assuming a timeout was set.
2. Workflow code in a `catch` block isn't always executed.
3. We see a "WorkflowTaskFailed" in the history with cause=unspecified. The failure details contain the same information as the failed activity.

## Steps to Reproduce the Problem
- Could not create a minimal reproducible example

## Specifications
  - Version:
  - Platform: 
  - SDK: TypeScript

![Screen Shot 2023-05-02 at 14 29 54](https://user-images.githubusercontent.com/26362206/235654331-7d7f615d-9996-499b-a929-7e502f43f0ba.png)

**Comment by bergundy** [comment]
The expected behavior that you're describing is something that we test extensively so I don't think it's an SDK bug.

My first goto would be to check if you have multiple versions of any `@temporalio/*` packages as we use `instanceof` checks to determine whether a workflow run or task should fail.

**Comment by Uria1** [comment]
I managed to solve the problem.
The issue was that activities that were started in parallel (without `await`) resulted in promises without proper error handling. I had to add a `catch` block to each activity promise, save any errors that are thrown, and then within the main workflow function, check if any errors occurred and fail the workflow by throwing an `ApplicationFailure`.

**Comment by bergundy** [comment]
@mjameswh we need to make unhandled rejection clearer. Will save a lot of headaches for users.

**Comment by mjameswh** [comment]
Duplicate of #1039

**Comment by Uria1** [comment]
> @mjameswh we need to make unhandled rejection clearer. Will save a lot of headaches for users.

Certainly. It wasn't clear and I waisted way too much time trying to fix it.
